File: leetcode/213_HouseRobber2.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0 ) return 0;
        if(n == 1) return nums[0];
        else if( n == 2) return Math.max(nums[0], nums[1]);
        return Math.max(robber(nums, 0, n-1), robber(nums, 1, n));
     
    }
    
    int robber(int []nums, int start, int end){
        int n = nums.length;
        int dp[] = new int[n];
        dp[0] = nums[0];
        dp[1] = nums[1];
        for(int i = start; i<end; i++){
            int max = Integer.MIN_VALUE;
            for( int j = start; j < i-1; j++){
                max = Math.max(max, dp[j]);
            }
            dp[i] = Math.max(nums[i], nums[i] + max);
        }
        
        int ans = Integer.MIN_VALUE;
        for(int i = start; i<end; i++)
        {
            ans = Math.max(ans, dp[i]);
        }
        return ans;
    }
    
}

File: leetcode/python/213_HouseRobber2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        elif n == 2:
            return max(nums[0], nums[1])

        amount = [0 for _ in range(n)]

        # assume first house is robbed 
        
        first = 0 
        amount[0] = nums[0]
        for i in range(1, n-1):
            amount[i] = max(first + nums[i], amount[i-1])
            first = amount[i-1] 

        max_amount = amount[n-2]
        # assume last house is robbed 
        last = 0
        amount[n-1] = nums[n-1]
        for i in range(n-2, 0, -1):
            amount[i] = max(last + nums[i], amount[i+1])
            last = amount[i+1]

        max_amount = max(max_amount, amount[1])
        return max_amount